File: lib/offline/bot/hand_evaluator.dart
import 'package:koutbh/shared/models/card.dart';
import 'package:koutbh/shared/logic/card_utils.dart';
import 'package:koutbh/offline/bot/bot_settings.dart';

enum PartnerAction { unknown, bid, passed }

class HandStrength {
  final double personalTricks;
  final Suit? strongestSuit;
  const HandStrength({required this.personalTricks, this.strongestSuit});
}

class HandEvaluator {
  /// Non-joker cards grouped by suit (empty suits omitted).
  static Map<Suit, List<GameCard>> suitDistribution(List<GameCard> hand) {
    final map = <Suit, List<GameCard>>{};
    for (final c in hand) {
      if (c.isJoker) continue;
      map.putIfAbsent(c.suit!, () => []).add(c);
    }
    return map;
  }

  /// Base trick probability for a card (no trump).
  static double _baseProbability(Rank rank) => switch (rank) {
    Rank.ace => 0.85,
    Rank.king => 0.65,
    Rank.queen => 0.35,
    Rank.jack => 0.15,
    _ => 0.05, // 10 and below
  };

  /// Bonus added when the card is in the prospective trump suit.
  static double _trumpBonus(Rank rank) => switch (rank) {
    Rank.ace => 0.15,
    Rank.king => 0.25,
    Rank.queen => 0.25,
    Rank.jack => 0.25,
    _ => 0.30, // 10 and below
  };

  static HandStrength evaluate(List<GameCard> hand) {
    if (hand.isEmpty) {
      return const HandStrength(personalTricks: 0.0);
    }

    final bySuit = suitDistribution(hand);
    final suitCounts = countBySuit(hand);

    // Step 1: Find strongest suit by raw trick potential (sum of base probs).
    final rawPotential = <Suit, double>{};
    for (final entry in bySuit.entries) {
      double sum = 0.0;
      for (final card in entry.value) {
        sum += _baseProbability(card.rank!);
      }
      rawPotential[entry.key] = sum;
    }

    Suit? strongest;
    double bestPotential = -1;
    for (final entry in rawPotential.entries) {
      if (entry.value > bestPotential) {
        bestPotential = entry.value;
        strongest = entry.key;
      }
    }

    // Step 2: Score each card with base + trump bonus for strongest suit.
    double score = 0.0;

    for (final card in hand) {
      if (card.isJoker) {
        score += 1.0; // Guaranteed trick
        continue;
      }

      final rank = card.rank!;
      final suit = card.suit!;
      double cardScore = _baseProbability(rank);
      if (suit == strongest) {
        cardScore += _trumpBonus(rank);
      }
      score += cardScore;
    }

    // Step 3: Suit texture bonuses.
    score += _suitTextureBonus(hand);

    // Step 4: Long suit bonus — +0.1 per card beyond 3 for suits with 4+.
    for (final entry in suitCounts.entries) {
      if (entry.value >= 4) {
        score += (entry.value - 3) * 0.1;
      }
    }

    // Step 5: Void bonuses.
    final hasTrump = strongest != null && (suitCounts[strongest] ?? 0) > 0;
    for (final suit in Suit.values) {
      if (suitCounts.containsKey(suit)) continue;
      // This suit is void.
      if (suit == strongest) continue; // Void in own trump: no bonus.
      if (hasTrump) {
        score += 1.0; // Ruffing potential
      } else {
        score += 0.1; // Void but no trump
      }
    }

    return HandStrength(
      personalTricks: score.clamp(0.0, 8.0),
      strongestSuit: strongest,
    );
  }

  /// Partner-adjusted effective tricks, clamped to 0.0-8.0.
  static double effectiveTricks(
    HandStrength strength, {
    required PartnerAction partnerAction,
  }) {
    final partnerEstimate = switch (partnerAction) {
      PartnerAction.unknown => BotSettings.partnerEstimateDefault,
      PartnerAction.bid => BotSettings.partnerEstimateBid,
      PartnerAction.passed => BotSettings.partnerEstimatePass,
    };
    return (strength.personalTricks + partnerEstimate).clamp(0.0, 8.0);
  }

  static double _suitTextureBonus(List<GameCard> hand) {
    double bonus = 0.0;
    final bySuit = <Suit, List<Rank>>{};
    for (final c in hand) {
      if (c.isJoker) continue;
      bySuit.putIfAbsent(c.suit!, () => []).add(c.rank!);
    }
    for (final ranks in bySuit.values) {
      final hasAce = ranks.contains(Rank.ace);
      final hasKing = ranks.contains(Rank.king);
      final hasQueen = ranks.contains(Rank.queen);
      if (hasAce && hasKing && hasQueen) {
        bonus += 0.5;
      } else if (hasAce && hasKing) {
        bonus += 0.3;
      } else if (hasKing && hasQueen && !hasAce) {
        bonus += 0.2;
      }
    }
    return bonus;
  }
}

File: workers/src/game/bot/hand-evaluator.ts
import { decodeCard } from '../card';
import type { SuitName, RankName } from '../types';

export interface HandStrength {
  personalTricks: number;
  strongestSuit: SuitName | null;
}

export type PartnerAction = 'unknown' | 'bid' | 'passed';

// Base trick probability for a card (no trump context).
function baseProbability(rank: RankName): number {
  switch (rank) {
    case 'ace': return 0.85;
    case 'king': return 0.65;
    case 'queen': return 0.35;
    case 'jack': return 0.15;
    default: return 0.05; // ten and below
  }
}

// Bonus added when the card is in the prospective trump suit.
function trumpBonus(rank: RankName): number {
  switch (rank) {
    case 'ace': return 0.15;
    case 'king': return 0.25;
    case 'queen': return 0.25;
    case 'jack': return 0.25;
    default: return 0.30; // ten and below
  }
}

function suitTextureBonus(bySuit: Map<SuitName, RankName[]>): number {
  let bonus = 0.0;
  for (const ranks of bySuit.values()) {
    const hasAce = ranks.includes('ace');
    const hasKing = ranks.includes('king');
    const hasQueen = ranks.includes('queen');
    if (hasAce && hasKing && hasQueen) bonus += 0.5;
    else if (hasAce && hasKing) bonus += 0.3;
    else if (hasKing && hasQueen && !hasAce) bonus += 0.2;
  }
  return bonus;
}

export function evaluateHand(hand: string[]): HandStrength {
  if (hand.length === 0) return { personalTricks: 0, strongestSuit: null };

  const bySuit = new Map<SuitName, RankName[]>();
  const suitCounts = new Map<SuitName, number>();
  let hasJoker = false;

  for (const code of hand) {
    const card = decodeCard(code);
    if (card.isJoker) { hasJoker = true; continue; }
    const suit = card.suit!;
    const rank = card.rank!;
    if (!bySuit.has(suit)) bySuit.set(suit, []);
    bySuit.get(suit)!.push(rank);
    suitCounts.set(suit, (suitCounts.get(suit) ?? 0) + 1);
  }

  let strongest: SuitName | null = null;
  let bestPotential = -1;
  for (const [suit, ranks] of bySuit) {
    const potential = ranks.reduce((s, r) => s + baseProbability(r), 0);
    if (potential > bestPotential) {
      bestPotential = potential;
      strongest = suit;
    }
  }

  let score = 0.0;

  if (hasJoker) score += 1.0; // guaranteed trick

  for (const [suit, ranks] of bySuit) {
    const isTrump = suit === strongest;
    for (const rank of ranks) {
      let cardScore = baseProbability(rank);
      if (isTrump) cardScore += trumpBonus(rank);
      score += cardScore;
    }
  }

  score += suitTextureBonus(bySuit);

  for (const count of suitCounts.values()) {
    if (count >= 4) score += (count - 3) * 0.1;
  }

  const hasTrump = strongest !== null && (suitCounts.get(strongest) ?? 0) > 0;
  for (const suit of ['spades', 'hearts', 'clubs', 'diamonds'] as SuitName[]) {
    if (bySuit.has(suit)) continue; // not void in this suit
    if (suit === strongest) continue; // void in own trump — no bonus
    score += hasTrump ? 1.0 : 0.1;
  }

  return {
    personalTricks: Math.max(0, Math.min(8, score)),
    strongestSuit: strongest,
  };
}

// Partner-contribution constants matching BotSettings in Dart.
const PARTNER_ESTIMATE_DEFAULT = 1.0;
const PARTNER_ESTIMATE_BID = 1.5;
const PARTNER_ESTIMATE_PASS = 0.5;

/** Partner-adjusted effective tricks, clamped to 0-8. */
export function effectiveTricks(
  strength: HandStrength,
  partnerAction: PartnerAction,
): number {
  const estimate = partnerAction === 'bid'
    ? PARTNER_ESTIMATE_BID
    : partnerAction === 'passed'
    ? PARTNER_ESTIMATE_PASS
    : PARTNER_ESTIMATE_DEFAULT;
  return Math.max(0, Math.min(8, strength.personalTricks + estimate));
}
